File: pdlearning/src/microservices/form-microservice/Api/Domain/Entities/FormParticipant.cs
using System;
using System.Diagnostics.CodeAnalysis;
using System.Linq.Expressions;
using Microservice.Form.Domain.ValueObjects.Form;
using Thunder.Platform.Core.Domain.Entities;

namespace Microservice.Form.Domain.Entities
{
    [SuppressMessage("Microsoft.Naming", "CA1724", Justification = "Toan Nguyen confirmed this.")]
    public class FormParticipant : BaseEntity, ISoftDelete
    {
        public Guid UserId { get; set; }

        public Guid FormOriginalObjectId { get; set; }

        public Guid FormId { get; set; }

        public DateTime AssignedDate { get; set; }

        public DateTime? SubmittedDate { get; set; }

        public bool? IsStarted { get; set; }

        public FormParticipantStatus Status { get; set; }

        public bool IsDeleted { get; set; }

        public static Expression<Func<FormParticipant, bool>> IsDoneExpr()
        {
            return x => x.SubmittedDate != null;
        }

        public static Expression<Func<FormParticipant, bool>> HasSaveAnswerPermissionExpr(Guid? userId)
        {
            return x => x.UserId == userId;
        }

        public bool HasSaveAnswerPermission(Guid? userId)
        {
            return HasSaveAnswerPermissionExpr(userId).Compile()(this);
        }
    }
}

File: pdlearning/src/web-apps/src/libraries/domain-api/src/form-participant/models/form-participant.ts
import { FormParticipantStatus } from './form-participant-status.enum';

export interface IFormParticipant {
  id?: string;
  formId?: string;
  userId?: string;
  assignedDate?: Date;
  submittedDate?: Date;
  status?: FormParticipantStatus;
  changedBy?: string;
  createdBy?: string;
  isStarted?: string;
}

export class FormParticipant implements IFormParticipant {
  public id?: string;
  public formId?: string;
  public userId?: string;
  public assignedDate?: Date;
  public submittedDate?: Date;
  public status?: FormParticipantStatus;
  public changedBy?: string;
  public createdBy?: string;
  public isStarted?: string;

  constructor(data?: IFormParticipant) {
    if (data == null) {
      return;
    }

    this.id = data.id;
    this.formId = data.formId;
    this.userId = data.userId;
    this.assignedDate = data.assignedDate;
    this.submittedDate = data.submittedDate;
    this.status = data.status;
    this.changedBy = data.changedBy;
    this.createdBy = data.createdBy;
    this.isStarted = data.isStarted;
  }
}
